receiver/qubes-psu-client.c

```diff
@@ -0,0 +1,196 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+#include <string.h>
+#include <stdbool.h>
+#include <signal.h>
+#include <json.h>
+
+#include "dummy_psu.h"
+
+#define BUFSIZE 4096
+
+int sigint_received = 0;
+int sigterm_received = 0;
+int cleaned_power_supplies = 0;
+
+struct psu {
+	int fd;
+	struct ioctl_psu specs;
+};
+
+void sig_handler(int s)
+{
+	if (s == SIGINT) {
+		fprintf(stderr, "INFO: SIGINT received\n");
+		sigint_received = 1;
+	}
+	if (s == SIGTERM) {
+		fprintf(stderr, "INFO: SIGTERM received\n");
+		sigterm_received = 1;
+	}
+}
+
+void clean_power_supplies(struct psu *power_supplies)
+{
+	int i;
+	for (i = 0; i < MAX_PSU; i++) {
+		if (power_supplies[i].fd >= 0) {
+			close(power_supplies[i].fd);
+		}
+	}
+	free(power_supplies);
+	cleaned_power_supplies = 1;
+}
+
+int main(int argc, char *argv[])
+{
+	signal(SIGINT, sig_handler);
+	signal(SIGTERM, sig_handler);
+
+	int i, ret = 0;
+	struct psu *power_supplies = calloc(MAX_PSU, sizeof(struct psu));
+	char buf[BUFSIZE];
+	struct json_object *jobj;
+	struct json_object *tmp;
+	char curr_type[MAX_KEYLENGTH];
+	char curr_name[MAX_KEYLENGTH];
+	int dev_num;
+	struct ioctl_pspval propval;
+	int status;
+	FILE *fp;
+	char cmd[79];
+
+	strncpy(cmd, "/usr/bin/qrexec-client-vm ", 27);
+
+	if (argc == 2) {
+		if (!strncmp("default", argv[1], 9)) {
+			strncat(cmd, "@default", 9);
+		} else {
+			strncat(cmd, argv[1], 33);
+		}
+	} else if (argc > 2) {
+		printf("ERROR: too many arguments provided\n");
+		return -1;
+	} else {
+		strncat(cmd, "@default", 9);
+	}
+
+	strncat(cmd, " qubes.PowerSupply", 19);
+
+	fp = popen(cmd, "r");
+	if (fp == NULL) {
+		printf("ERROR: failed to open pipe\n");
+		return -1;
+	}
+
+	while (fgets(buf, BUFSIZE, fp) != NULL) {
+		if (sigint_received || sigterm_received) {
+			break;
+		}
+		setvbuf(stdout, NULL, _IONBF, 0);
+		jobj = json_tokener_parse(buf);
+
+		json_object_object_get_ex(jobj, "NAME", &tmp);
+		strncpy(curr_name, json_object_get_string(tmp), MAX_KEYLENGTH - 1);
+
+		json_object_object_get_ex(jobj, "TYPE", &tmp);
+		strncpy(curr_type, json_object_get_string(tmp), MAX_KEYLENGTH - 1);
+
+		dev_num = -1;
+		if (strcmp(curr_name, "\0") && strcmp(curr_name, "\0")) {
```

Comment:
This is the same as `strcmp(..., "")`, and also - the second one should be curr_type.